File: patterns/hash-map/242. Valid Anagram/242. Valid Anagram.js
/**
 * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
 *
 * An Anagram is a word or phrase formed by rearranging the letters of a different
 * word or phrase, typically using all the original letters exactly once.
 */
var isAnagram = function (s, t) {
    if (s.length != t.length) return false;

    hmap = {};
    for (let x of s) {
        if (hmap[x]) {
            hmap[x] = hmap[x] + 1;
        } else {
            hmap[x] = 1;
        }
    }
    for (let x of t) {
        if (hmap[x]) {
            hmap[x] = hmap[x] - 1;
        }
    }
    let flag = true;
    for (let x in hmap) {
        if (hmap[x] !== 0) {
            return false;
        }
    }
    return flag;
};
console.log(isAnagram("rat", "car")); // false

/*
Time complexity: O(N)
Space complexity: O(1) - since we use fixed size array for character counting
where N is the length of the input strings.
*/

File: patterns/hash-map/242. Valid Anagram/242. Valid Anagram.py
class Solution:
    def isAnagram(s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        hmap = {}

        # count chars from s
        for ch in s:
            if ch in hmap:
                hmap[ch] += 1
            else:
                hmap[ch] = 1

        # reduce counts with t
        for ch in t:
            if ch in hmap:
                hmap[ch] -= 1

        # check all counts
        for val in hmap.values():
            if val != 0:
                return False

        return True


s = Solution()
print(s.isAnagram("aacc", "ccac"))

"""
Time complexity: O(N)
Space complexity: O(1) - since we use fixed size array for character counting
where N is the length of the input strings.
"""

"""
Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Input: s = "anagram", t = "nagaram"
Output: true
"""
